### PR TITLE
🏃 Use controller-runtime cache for watching remote nodes over informers

### LINKED ISSUE #2412: Watching node in remote clusters should use a cache
**Note: this is only applicable once #2250 is merged**

Watching of nodes in remote clusters was introduced as part of the Machine Health Checking work in #2250.

Currently it uses [individual node informers](https://github.com/kubernetes-sigs/cluster-api/blob/79cf0941adbed5f92d9f2829a428162f9bc6dc25/controllers/machinehealthcheck_controller.go#L367-L370) from client-go to subscribe to events for nodes within the remote clusters, however, it could use the newly merged controller-runtime source type for `kindWithCache` which is designed for this purpose (https://github.com/kubernetes-sigs/controller-runtime/pull/794) which would make the implementation more controller-runtime native and a bit simpler too.

Once https://github.com/kubernetes-sigs/controller-runtime/pull/794 is available in a release, the dependency on controller-runtime should be updated and this should be implemented

/assign @JoelSpeed

### PR DESCRIPTION
**Pull request by JoelSpeed**
**What this PR does / why we need it**:

As per [suggestion in #2250](https://github.com/kubernetes-sigs/cluster-api/pull/2250/files#r382669297), the watching of remote nodes could be improved by the introduction of the a `KindWithCache` source from controller-runtime. This feature is not yet released however. Once the next release of controller-runtime is made, and the project is happy to bump the dependency, this PR can be updated and the improvements merged

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2412

### CODE REVIEW

**Comment by JoelSpeed** [comment]
I've manually tested this change doesn't break the remote watching of nodes, could see reconciliation occurring at the same timestamp as the node condition `lastTransition` occurred, so I think this is safe, just need to check the code over and wait for a CR release

File: controllers/machinehealthcheck_controller.go [hunk 1]
```diff
@@ -333,20 +330,20 @@ func (r *MachineHealthCheckReconciler) watchClusterNodes(ctx context.Context, c
 	return r.createClusterNodeInformer(ctx, c, key)
 }
 
-func (r *MachineHealthCheckReconciler) getClusterNodeInformer(key client.ObjectKey) (cache.Informer, bool) {
-	r.clusterNodeInformersLock.RLock()
-	defer r.clusterNodeInformersLock.RUnlock()
+func (r *MachineHealthCheckReconciler) getClusterNodeInformer(key client.ObjectKey) (cache.Cache, bool) {
```
Comment (detiber) [inline controllers/machinehealthcheck_controller.go]:
```suggestion
func (r *MachineHealthCheckReconciler) getClusterNodeCache(key client.ObjectKey) (cache.Cache, bool) {
```

File: controllers/machinehealthcheck_controller.go [hunk 2]
```diff
@@ -333,20 +330,20 @@ func (r *MachineHealthCheckReconciler) watchClusterNodes(ctx context.Context, c
...
+	r.clusterCachesLock.RLock()
+	defer r.clusterCachesLock.RUnlock()
 
-	informer, ok := r.clusterNodeInformers[key]
-	return informer, ok
+	c, ok := r.clusterCaches[key]
+	return c, ok
 }
 
 func (r *MachineHealthCheckReconciler) createClusterNodeInformer(ctx context.Context, c client.Client, key client.ObjectKey) error {
```
Comment (detiber) [inline controllers/machinehealthcheck_controller.go]:
```suggestion
func (r *MachineHealthCheckReconciler) createClusterNodeCache(ctx context.Context, c client.Client, key client.ObjectKey) error {
```

File: controllers/machinehealthcheck_controller.go (hunk 1 above)
Comment (JoelSpeed) [reply controllers/machinehealthcheck_controller.go]: Good spot! This is why I shouldn't rebase code on a Monday morning 😪 ☕️ 😂 Thanks @detiber

File: controllers/machinehealthcheck_controller.go [hunk 3]
```diff
@@ -356,25 +353,21 @@ func (r *MachineHealthCheckReconciler) createClusterNodeInformer(ctx context.Con
...
 	if err != nil {
-		return errors.Wrap(err, "error constructing remote cluster client")
+		return errors.Wrap(err, "error creating cache for remote cluster")
 	}
-
-	// TODO(JoelSpeed): See if we use the resync period from the manager instead of 0
-	factory := informers.NewSharedInformerFactory(k8sClient, 0)
-	nodeInformer := factory.Core().V1().Nodes().Informer()
-	go nodeInformer.Run(ctx.Done())
+	go clusterCache.Start(ctx.Done())
```
Comment (ncdc) [inline controllers/machinehealthcheck_controller.go]: xref #2577: I think we're going to need to able to stop a cache when the associated Cluster is deleted. This probably means creating a stop channel per cache and tracking it in the map, and adding logic that closes the stop channel for a cache if its corresponding Cluster is not found when reconciling.

**Comment by vincepri** [comment]
@JoelSpeed https://github.com/kubernetes-sigs/controller-runtime/releases new controller runtime release is up

**Comment by JoelSpeed** [comment]
Thanks @vincepri for the ping. I've updated the PR to update the dependencies to the new patch release. Put this as a separate commit for now, wasn't sure whether it was ok to just merge in with this of if it should be separate for some reason?

**Comment by vincepri** [comment]
@ncdc did you want to take a final look here?

**Comment by ncdc** [comment]
Sure, on it now

**Comment by vincepri** [comment]
/retitle 🏃 Use controller-runtime cache for watching remote nodes over informers

**Pull request merged**